### PR TITLE
add l1 loss op in python

### PR DESCRIPTION
**Pull request by MARD1NO**
增加 L1 Loss op

- [x] 增加后向梯度的验证

### CODE REVIEW

File: oneflow/python/test/ops/test_l1_loss.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+    flow.clear_default_session()
+    flow.env.init()
+    if device_type == "cpu":
+        flow.config.cpu_device_num(device_counts)
+    else:
+        flow.config.gpu_device_num(device_counts)
+
+    func_config = flow.FunctionConfig()
+
+    @flow.global_function(function_config=func_config, type="train")
```
Comment (doombeaker) [inline oneflow/python/test/ops/test_l1_loss.py:41]: type  和 function_conf 按照test case的编程习惯，type写前面。

File: oneflow/python/test/ops/test_l1_loss.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+"""
+import oneflow as flow
+import numpy as np
+import oneflow.typing as tp
+from test_util import GenArgList
+import unittest
+from collections import OrderedDict
+
+
+def compare_l1loss_with_np(
```
Comment (doombeaker) [inline oneflow/python/test/ops/test_l1_loss.py:24]: 仅当前脚本使用的函数，习惯上下划线开头

File: oneflow/python/test/ops/test_l1_loss.py [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+        arg_dict = OrderedDict()
+        arg_dict["input_shape"] = [(3, 16, 16)]
+        arg_dict["target_shape"] = [(3, 16, 16)]
+        arg_dict["device_type"] = ["cpu"]
+        arg_dict["machine_ids"] = ["0:0"]
+        arg_dict["device_counts"] = [1]
+        for arg in GenArgList(arg_dict):
+            _compare_l1loss_with_np(*arg)
+
+    def test_l1loss_gpu(test_case):
```
Comment (doombeaker) [inline oneflow/python/test/ops/test_l1_loss.py:102]: GPU 的test case 要加上 `@unittest.skipIf(os.getenv("ONEFLOW_TEST_CPU_ONLY"), "only test cpu cases")`，要不然过不了CPU-ONLY的测试。可以参考 `test_broadcast_maximum.py` 等文件

File: oneflow/python/test/ops/test_l1_loss.py [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+        np_l1_sum = np.sum(np_l1)
+
+        return [np_l1, np_l1_mean, np_l1_sum]
+
+    of_out_l1loss = oneflow_l1loss(input, target)
+    np_out_l1loss = np_l1loss(input, target)
+
+    assert np.allclose(of_out_l1loss[0], np_out_l1loss[0])
+
+    for i in range(1, len(np_out_l1loss)):
```
Comment (doombeaker) [inline oneflow/python/test/ops/test_l1_loss.py:84]: 这里是不是把numpy的标量转成数字，再直接用 numpy.allcose 更好？循环一个一个比效率低也不好看。

File: oneflow/python/test/ops/test_l1_loss.py (hunk 3 above)
Comment (doombeaker) [inline oneflow/python/test/ops/test_l1_loss.py:102]: GPU 的test case 要加上 `@unittest.skipIf(os.getenv("ONEFLOW_TEST_CPU_ONLY"), "only test cpu cases")`，要不然过不了CPU-ONLY的测试。可以参考 `test_broadcast_maximum.py` 等文件

File: oneflow/python/test/ops/test_l1_loss.py [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+
+    @flow.global_function(type="train", function_config=func_config)
+    def oneflow_l1loss(
+        of_input: tp.Numpy.Placeholder(shape=input.shape),
+        of_target: tp.Numpy.Placeholder(shape=target.shape),
+    ) -> List[tp.Numpy]:
+        x_var = flow.get_variable(
+            shape=input.shape,
+            dtype=flow.float32,
+            initializer=flow.ones_initializer(),
```
Comment (doombeaker) [inline oneflow/python/test/ops/test_l1_loss.py:50]: 这里用 `ones_initializer` 是有什么特别用意吗？一般用 `zeros_initializer` 更直观，不会让其它人多想？

File: oneflow/python/ops/nn_ops.py [hunk 6]
```diff
@@ -2646,3 +2646,100 @@ def leaky_relu_Job(x: tp.Numpy.Placeholder((5, ),)
...
+    reduction: str = "mean",
+    name: Optional[str] = None,
+) -> remote_blob_util.BlobDef:
+    """This operator computes the L1 Loss between each element in `input` and `target`. 
+
+    The equation is: 
+
+    .. math:: 
+
+        output = |Target - Input|
```
Comment (doombeaker) [inline oneflow/python/ops/nn_ops.py:2664]: 还没有添加 reduction 参数有关的说明以及公式变化

File: oneflow/python/test/ops/test_l1_loss.py [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+import oneflow as flow
+import numpy as np
+import oneflow.typing as tp
+from test_util import GenArgList
+import unittest
+from collections import OrderedDict
+from typing import List
+
+
+def _compare_l1loss_with_np(
```
Comment (doombeaker) [inline oneflow/python/test/ops/test_l1_loss.py:25]: 需要添加关于反向diff的测试

**Review by doombeaker — CHANGES_REQUESTED** [review]
添加反向diff的测试，以及其它comment中相关

File: oneflow/python/test/ops/test_l1_loss.py (hunk 4 above)
Comment (MARD1NO) [reply oneflow/python/test/ops/test_l1_loss.py:84]: 呃这里我是拿一个job做了 reduce='none', 'mean', 'sum' 的三组测试。返回一个里面有3个元素的列表。 

numpy也是这样做

然后列表每个元素进行比较。

File: oneflow/python/test/ops/test_l1_loss.py (hunk 5 above)
Comment (MARD1NO) [reply oneflow/python/test/ops/test_l1_loss.py:50]: 只是我随便加的，换成zero有啥区别吗，是为了减少对代码的误解？

File: oneflow/python/ops/nn_ops.py (hunk 6 above)
Comment (MARD1NO) [reply oneflow/python/ops/nn_ops.py:2664]: 好的，在mse那里加了这里忘记加了，后续会补上

File: oneflow/python/test/ops/test_l1_loss.py (hunk 7 above)
Comment (MARD1NO) [reply oneflow/python/test/ops/test_l1_loss.py:25]: 好的，添加到todo list

File: oneflow/python/ops/nn_ops.py [hunk 8]
```diff
@@ -2646,3 +2646,100 @@ def leaky_relu_Job(x: tp.Numpy.Placeholder((5, ),)
...
+
+        import oneflow as flow
+        import oneflow.typing as tp
+        import numpy as np
+
+
+        @flow.global_function()
+        def l1_job(x: tp.Numpy.Placeholder(shape=(3, 3)),
+                y: tp.Numpy.Placeholder(shape=(3, 3))) -> tp.Numpy:
+            # It's equal to `Torch.nn.L1Loss(reduction="mean")
```
Comment (wyg1997) [inline oneflow/python/ops/nn_ops.py:2689]: 我们是不是不需要专门标出来对标其它哪个框架的哪个op？

File: oneflow/python/ops/nn_ops.py [hunk 9]
```diff
@@ -2646,3 +2646,100 @@ def leaky_relu_Job(x: tp.Numpy.Placeholder((5, ),)
...
+
+        import oneflow as flow
+        import oneflow.typing as tp
+        import numpy as np
+
+
+        @flow.global_function()
+        def l1_job(x: tp.Numpy.Placeholder(shape=(3, 3)),
+                y: tp.Numpy.Placeholder(shape=(3, 3))) -> tp.Numpy:
+            # It's equal to `Torch.nn.L1Loss(reduction="sum")
```
Comment (wyg1997) [inline oneflow/python/ops/nn_ops.py:2714]: 这个也一样，不需要专门标出

File: oneflow/python/ops/nn_ops.py [hunk 10]
```diff
@@ -2646,3 +2646,100 @@ def leaky_relu_Job(x: tp.Numpy.Placeholder((5, ),)
...
+
+        # output [24.]
+
+    """
+    assert (
+        input.shape == target.shape
+    ), "The Input shape must be the same as Target shape"
+
+    if name is None:
+        name = "L1Loss"
```
Comment (wyg1997) [inline oneflow/python/ops/nn_ops.py:2733]:
```suggestion
        name = id_util.UniqueStr("L1Loss_")
```

File: oneflow/python/ops/nn_ops.py [hunk 11]
```diff
@@ -2646,3 +2646,100 @@ def leaky_relu_Job(x: tp.Numpy.Placeholder((5, ),)
...
+    The equation is: 
+
+    .. math:: 
+
+        output = |Target - Input|
+
+    Args:
+        input (remote_blob_util.BlobDef): The input Blob.  
+        target (remote_blob_util.BlobDef): The target value. 
+        reduction (str): The reduce type, it can be the one of "none", "mean", "sum". Defaults to "mean".
```
Comment (wyg1997) [inline oneflow/python/ops/nn_ops.py]: 代码中没有对 `reduction` 做检察。

Comment (MARD1NO) [reply oneflow/python/ops/nn_ops.py]: 谢谢提醒，新版本已增加 T T

File: oneflow/python/ops/nn_ops.py [hunk 12]
```diff
@@ -2646,3 +2646,120 @@ def leaky_relu_Job(x: tp.Numpy.Placeholder((5, ),)
         .InferAndTryRun()
         .RemoteBlobList()[0]
     )
+
```
Comment (doombeaker) [inline oneflow/python/ops/nn_ops.py:2649]: 还有个 文件 `oneflow/python/ops/loss_ops.py`，好像是专门放loss的，但是目前只放了一个 smooth_loss，你可以权衡下，放哪个文件比较好。

File: oneflow/python/test/ops/test_l1_loss.py [hunk 13]
```diff
@@ -0,0 +1,184 @@
+"""
```
Comment (doombeaker) [inline oneflow/python/test/ops/test_l1_loss.py:1]: 文件名叫做 `test_l1loss.py` 更好？现在的习惯都是 `test_opname.py`，你的op是 `l1loss`不是`l1_loss`

File: oneflow/python/test/ops/test_l1_loss.py [hunk 14]
```diff
@@ -0,0 +1,184 @@
...
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+import oneflow as flow
+import numpy as np
+import oneflow.typing as tp
+from test_util import GenArgList
+import unittest
+from collections import OrderedDict
+from typing import List, Dict
```
Comment (doombeaker) [inline oneflow/python/test/ops/test_l1_loss.py:22]: List 已经没有用到了

File: oneflow/python/test/ops/test_l1_loss.py [hunk 15]
```diff
@@ -0,0 +1,184 @@
...
+            "np_l1_loss_sum": np_l1_sum,
+        }
+
+        return np_l1_dict
+
+    def np_l1_loss_diff(np_input, np_target):
+        # Use numpy to compute diff
+        original_shape = np_target.shape
+        elemcnt = np_target.size
+        elemcnt = np.reshape(elemcnt, -1)
```
Comment (doombeaker) [inline oneflow/python/test/ops/test_l1_loss.py:63]: 这句多余了

**Review by doombeaker — CHANGES_REQUESTED** [review]
考虑下文件名，去除不必要的代码

File: oneflow/python/ops/nn_ops.py [hunk 16]
```diff
@@ -2646,3 +2646,120 @@ def leaky_relu_Job(x: tp.Numpy.Placeholder((5, ),)
...
+    assert reduction in [
+        "none",
+        "mean",
+        "sum",
+    ], "{} is not a valid value for reduction, The reduction must be the one of `none`, `mean`, `sum`. ".format(
+        reduction
+    )
+
+    if name is None:
+        name = "L1Loss_"
```
Comment (doombeaker) [inline oneflow/python/ops/nn_ops.py:2753]: 这个 `_` 是多余的？因为你后面做字符串拼接已经有下划线。
虽然不影响功能，但是和你写的另外的loss的风格不一致

File: oneflow/python/ops/nn_ops.py (hunk 12 above)
Comment (MARD1NO) [reply oneflow/python/ops/nn_ops.py:2649]: 我有想过。但如果和torch对齐，是放在nn模块下的话（nn.L1Loss)

在 loss_ops.py下，export成 `nn.L1Loss` 会不会有点怪

**Review by doombeaker — APPROVED** [review]
